### ISSUE TITLE
package-helper.el issues

### ISSUE THREAD

**Issue by ustcscgy** [skeeto/.emacs.d#3]
1. The link in http://nullprogram.com/blog/2013/06/02/ is broken.

https://github.com/skeeto/.emacs.d/blob/master/etc/package-helper.el
to
https://github.com/skeeto/.emacs.d/blob/master/lisp/package-helper.el

2. The default behavior of using package.el and and install if the package is not "installed" causes trouble in my situation. I use el-get instead of package.el to manage package installation. 

I think it's better to separate installation and configuration.

**Comment by ustcscgy** [comment]
I'm going to close this issue, before that, I'd like to discuss `use-package` and `with-package`:

You have chose to use `use-package` instead of `with-package`, however, I think `with-package` still have some advantages:
1. `use-package :config` does not use `eval-after-load` by default.
2. When `eval-after-load` is used, the argument passed to it is a quote form and can not be compiled (unlike `with-package`).
3. `with-package` can have multiple packages as its first argument, and the customize code runs only once.

By the way, a dozen keywords of `use-package` seems too complicated and unnecessary to me. I consider `with-package` does a better job of "modularize package configuration, tidy and fast". What's your opinion?

**Comment by skeeto** [comment]
> use-package :config does not use eval-after-load by default.

It does the right thing when deferred — either explicitly as `:defer t` or implicitely).

~~~el
(use-package foo
  :defer t
  :config (foobar))

;; macroexpand-all =>
(eval-after-load 'foo
  #'(lambda nil
      (prog1 t
        (condition-case err
            (foobar)
          ((debug error)
           (ignore
            (display-warning 'use-package
                             (format "%s %s: %s" "foo" ":config"
                                     (error-message-string err))
                             :error)))))))
~~~

The sharp-quote allows the function to be compiled properly.

> with-package can have multiple packages as its first argument, and the customize code runs only once.

Yup, this is an advantage of with-package. I have `(require ...)` in some of my `:config` to pull in related packages. However, it doesn't really bother me and perhaps the simpler approach is better anyway.

> By the way, a dozen keywords of use-package seems too complicated and unnecessary to me. I consider with-package does a better job of "modularize package configuration, tidy and fast". What's your opinion?

I definitely like use-package better. First, I don't have to maintain it. :-) Second, with-package's syntax is more concise but now-a-days I generally prefer more explicit. A number of times I had to re-learn with-package but use-package remains obvious (except for the rules on :defer, so I always make it explicit).

**Comment by ustcscgy** [comment]
Thanks for your reply.
